File: src/dataset/javascript/linked-list_exercism-ThomasZumsteg/gemini/GeminiFlash_linked-list_v3.js
var LinkedList = function() {
	this.head = null;
	this.tail = null;
	this.length = 0;
};

LinkedList.prototype.push = function(val) {
	const newElement = { val: val, next: null, prev: this.tail };

	if (this.tail) {
		this.tail.next = newElement;
	} else {
		this.head = newElement;
	}

	this.tail = newElement;
	this.length++;
};

LinkedList.prototype.pop = function() {
	if (this.length === 0) {
		return undefined;
	}

	const element = this.tail;
	const val = element.val;

	this.tail = element.prev;

	if (this.tail) {
		this.tail.next = null;
	} else {
		this.head = null;
	}

	this.length--;
	return val;
};

LinkedList.prototype.unshift = function(val) {
	const newElement = { val: val, next: this.head, prev: null };

	if (this.head) {
		this.head.prev = newElement;
	} else {
		this.tail = newElement;
	}

	this.head = newElement;
	this.length++;
};

LinkedList.prototype.shift = function() {
	if (this.length === 0) {
		return undefined;
	}

	const element = this.head;
	const val = element.val;

	this.head = element.next;

	if (this.head) {
		this.head.prev = null;
	} else {
		this.tail = null;
	}

	this.length--;
	return val;
};

LinkedList.prototype.count = function() {
	return this.length;
};

LinkedList.prototype.delete = function(val) {
	let element = this.head;

	while (element && element.val !== val) {
		element = element.next;
	}

	if (element) {
		if (element.prev) {
			element.prev.next = element.next;
		} else {
			this.head = element.next;
		}

		if (element.next) {
			element.next.prev = element.prev;
		} else {
			this.tail = element.prev;
		}

		this.length--;
	}
};

module.exports = LinkedList;

File: src/dataset/python/linked-list_exercism-ThomasZumsteg/gemini/GeminiFlash_linked_list_v3.py
class Node(object):
    def __init__(self, value, next=None, previous=None):
        self.value = value
        self.next = next
        self.previous = previous


class LinkedList(object):
    def __init__(self):
        self.head = None
        self._len = 0

    def __len__(self):
        return self._len

    def __iter__(self):
        if self.head is None:
            return

        current = self.head
        yield current.value
        current = current.next

        while current != self.head:
            yield current.value
            current = current.next

    def push(self, value):
        new_node = Node(value)
        if self.head is None:
            new_node.next = new_node.previous = new_node
            self.head = new_node
        else:
            tail = self.head.previous
            new_node.next = self.head
            new_node.previous = tail
            self.head.previous = new_node
            tail.next = new_node
            self.head = new_node
        self._len += 1

    def pop(self):
        if self.head is None:
            raise IndexError("pop from an empty LinkedList")

        node = self.head
        value = node.value

        if self.head.next == self.head:
            self.head = None
        else:
            tail = self.head.previous
            self.head = node.next
            tail.next = self.head
            self.head.previous = tail

        self._len -= 1
        return value

    def shift(self):
        if self.head is None:
            raise IndexError("shift from an empty LinkedList")

        value = self.head.previous.value
        if self.head.next == self.head:
            self.head = None
        else:
            new_tail = self.head.previous.previous
            new_tail.next = self.head
            self.head.previous = new_tail
        self._len -= 1
        return value

    def unshift(self, value):
        new_node = Node(value)
        if self.head is None:
            new_node.next = new_node.previous = new_node
            self.head = new_node
        else:
            tail = self.head.previous
            new_node.next = self.head
            new_node.previous = tail
            self.head.previous = new_node
            tail.next = new_node
            self.head = new_node
        self._len += 1